### PR TITLE
Port HTTP/2 article from HPBN 

### LINKED ISSUE #3464: Introduction to HTTP/2 guide
Existing content: 
https://surma.link/things/h2setup/
http://blog.chromium.org/2015/02/hello-http2-goodbye-spdy.html
http://blog.chromium.org/2016/02/transitioning-from-spdy-to-http2.html
https://docs.google.com/presentation/d/1r7QXGYOLCh4fcUq0jDdDwKJWNqWK1o4xMtYpKZCJYjM/edit#slide=id.p19
hpbn.co/http2
https://docs.google.com/document/d/1K0NykTXBbbbTlv60t5MyJvXjqKGsCVNYHyLEXIxYMv0/edit (H/2 Push specific)

Partially related to #3459 but also just to have a good one-pager on it for /web.

### PR DESCRIPTION
**Pull request by surma**
Closes #3464

This is almost a full copy of @igrigorik article on HPBN. I mostly stripped the very low-level parts at the bottom.

@Meggin @addyosmani PTAL

### CODE REVIEW

File: src/content/en/fundamentals/performance/http2/index.md [hunk 1]
```diff
@@ -0,0 +1,236 @@
+project_path: /web/_project.yaml
+book_path: /web/fundamentals/_book.yaml
+description: HTTP/2 will make our applications faster, simpler, and more robust
+
+{# wf_updated_on: 2016-09-29 #}
+{# wf_published_on: 2016-09-29 #}
+
+# Introduction to HTTP/2
```
Comment (petele) [inline src/content/en/fundamentals/performance/http2/index.md]: Missing ` {: .page-title }`, which is causing the build to fail. All pages need a page title. Should be in the format of: `# Introduction to HTTP/2  {: .page-title }`

File: src/content/en/fundamentals/performance/http2/index.md [hunk 2]
```diff
@@ -0,0 +1,236 @@
+project_path: /web/_project.yaml
+book_path: /web/fundamentals/_book.yaml
+description: HTTP/2 will make our applications faster, simpler, and more robust
```
Comment (petele) [inline src/content/en/fundamentals/performance/http2/index.md]: this is the meta description, mostly meant for search, I'd recommend something more along the lines of what people would search for.

File: src/content/en/fundamentals/performance/http2/index.md [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+{# wf_updated_on: 2016-09-29 #}
+{# wf_published_on: 2016-09-29 #}
+
+# Introduction to HTTP/2
+
+{% include "web/_shared/contributors/surma.html" %}
+
+HTTP/2 will make our applications faster, simpler, and more robust — a rare combination — by allowing us to undo many of the HTTP/1.1 workarounds previously done within our applications and address these concerns within the transport layer itself. Even better, it also opens up a number of entirely new opportunities to optimize our applications and improve performance!
```
Comment (petele) [inline src/content/en/fundamentals/performance/http2/index.md]: Please break lines at 80 chars

File: src/content/en/fundamentals/performance/http2/index.md [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+SPDY was an experimental protocol, developed at Google and announced in mid-2009, whose primary goal was to try to reduce the load latency of web pages by addressing some of the well-known performance limitations of HTTP/1.1. Specifically, the outlined project goals were set as follows:
+
+* Target a 50% reduction in page load time (PLT).
+* Avoid the need for any changes to content by website authors.
+* Minimize deployment complexity, avoid changes in network infrastructure.
+* Develop this new protocol in partnership with the open-source community.
+* Gather real performance data to (in)validate the experimental protocol.
+
+> Note: To achieve the 50% PLT improvement, SPDY aimed to make more efficient use of the underlying TCP connection by introducing a new binary framing layer to enable request and response multiplexing, prioritization, and header compression; see [Latency as a Performance Bottleneck](https://hpbn.co/primer-on-web-performance/#latency-as-a-performance-bottleneck).
```
Comment (petele) [inline src/content/en/fundamentals/performance/http2/index.md]: To get the Note syntax, just use `Note: blah` Don't wrap it in a blockquote

File: src/content/en/fundamentals/performance/http2/index.md [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+
+Not long after the initial announcement, Mike Belshe and Roberto Peon, both software engineers at Google, shared their first results, documentation, and source code for the experimental implementation of the new SPDY protocol:
+
+> So far we have only tested SPDY in lab conditions. The initial results are very encouraging: when we download the top 25 websites over simulated home network connections, we see a significant improvement in performance—pages loaded up to 55% faster. *(Chromium Blog)*
```
Comment (petele) [inline src/content/en/fundamentals/performance/http2/index.md]: Link to post if possible

File: src/content/en/fundamentals/performance/http2/index.md [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+
+Observing the above trend, the HTTP Working Group (HTTP-WG) kicked off a new effort to take the lessons learned from SPDY, build and improve on them, and deliver an official "HTTP/2" standard: a new charter was drafted, an open call for HTTP/2 proposals was made, and after a lot of discussion within the working group, the SPDY specification was adopted as a starting point for the new HTTP/2 protocol.
+
+Over the next few years SPDY and HTTP/2 would continue to coevolve in parallel, with SPDY acting as an experimental branch that was used to test new features and proposals for the HTTP/2 standard: what looks good on paper may not work in practice, and vice versa, and SPDY offered a route to test and evaluate each proposal before its inclusion in the HTTP/2 standard. In the end, this process spanned three years and resulted in a over a dozen intermediate drafts:
+
+* March 2012: Call for proposals for HTTP/2
```
Comment (petele) [inline src/content/en/fundamentals/performance/http2/index.md]: Does it make sense to link to any of these?

File: src/content/en/fundamentals/performance/http2/index.md [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+* August 2014: Working Group last call for HTTP/2
+* February 2015: IESG approved HTTP/2 and HPACK drafts
+* May 2015: RFC 7540 (HTTP/2) and RFC 7541 (HPACK) are published
+
+In early 2015 the IESG reviewed and approved the new HTTP/2 standard for publication. Shortly after that, the Google Chrome team announced their schedule to deprecate SPDY and NPN extension for TLS:
+
+> HTTP/2's primary changes from HTTP/1.1 focus on improved performance. Some key features such as multiplexing, header compression, prioritization and protocol negotiation evolved from work done in an earlier open, but non-standard protocol named SPDY. Chrome has supported SPDY since Chrome 6, but since most of the benefits are present in HTTP/2, it’s time to say goodbye. We plan to remove support for SPDY in early 2016, and to also remove support for the TLS extension named NPN in favor of ALPN in Chrome at the same time. Server developers are strongly encouraged to move to HTTP/2 and ALPN.
+>
+> We’re happy to have contributed to the open standards process that led to HTTP/2, and hope to see wide adoption given the broad industry engagement on standardization and implementation. *(Chromium Blog)*
```
Comment (petele) [inline src/content/en/fundamentals/performance/http2/index.md]: Link to post if possible.

File: src/content/en/fundamentals/performance/http2/index.md [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+
+As the above examples illustrate, the combination of stream dependencies and weights provides an expressive language for resource prioritization, which is a critical feature for improving browsing performance where we have many resource types with different dependencies and weights. Even better, the HTTP/2 protocol also allows the client to update these preferences at any point, which enables further optimizations in the browser—e.g., we can change dependencies and reallocate weights in response to user interaction and other signals.
+
+> Note: Stream dependencies and weights express a transport preference, not a requirement, and as such do not guarantee a particular processing or transmission order. That is, the client cannot force the server to process the stream in particular order using stream prioritization. While this may seem counterintuitive, it is in fact the desired behavior: we do not want to block the server from making progress on a lower priority resource if a higher priority resource is blocked.
```
Comment (petele) [inline src/content/en/fundamentals/performance/http2/index.md]: Use standard note syntax here

File: src/content/en/fundamentals/performance/http2/index.md [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+
+## One Connection Per Origin
+
+With the new binary framing mechanism in place, HTTP/2 no longer needs multiple TCP connections to multiplex streams in parallel; each stream is split into many frames, which can be interleaved and prioritized. As a result, all HTTP/2 connections are persistent, and only one connection per origin is required, which offers numerous performance benefits.
+
+> For both SPDY and HTTP/2 the killer feature is arbitrary multiplexing on a single well congestion controlled channel. It amazes me how important this is and how well it works. One great metric around that which I enjoy is the fraction of connections created that carry just a single HTTP transaction (and thus make that transaction bear all the overhead). For HTTP/1 74% of our active connections carry just a single transaction—persistent connections just aren’t as helpful as we all want. But in HTTP/2 that number plummets to 25%. That’s a huge win for overhead reduction. *(HTTP/2 is Live in Firefox, Patrick McManus)*
```
Comment (petele) [inline src/content/en/fundamentals/performance/http2/index.md]: Link to quote if possible

File: src/content/en/fundamentals/performance/http2/index.md [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+
+Most HTTP transfers are short and bursty, whereas TCP is optimized for long-lived, bulk data transfers. By reusing the same connection HTTP/2 is able to both make more efficient use of each TCP connection, and also significantly reduce the overall protocol overhead. Further, the use of fewer connections reduces the memory and processing footprint along the full connection path (i.e., client, intermediaries, and origin servers), which reduces the overall operational costs and improves network utilization and capacity. As a result, the move to HTTP/2 should not only reduce the network latency, but also help improve throughput and reduce the operational costs.
+
+> Note: Reduced number of connections is a particularly important feature for improving performance of HTTPS deployments: this translates to fewer expensive TLS handshakes, better session reuse, and an overall reduction in required client and server resources.
```
Comment (petele) [inline src/content/en/fundamentals/performance/http2/index.md]: Use standard note syntax here

File: src/content/en/fundamentals/performance/http2/index.md [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+> On the lower-bandwidth DSL link, in which the upload link is only 375 Kbps, request header compression in particular, led to significant page load time improvements for certain sites (i.e., those that issued large number of resource requests). We found a reduction of 45–1142 ms in page load time simply due to header compression. *(SPDY whitepaper, chromium.org)*
+
+However, in the summer of 2012, a "CRIME" security attack was published against TLS and SPDY compression algorithms, which could result in session hijacking. As a result, the zlib compression algorithm was replaced by HPACK, which was specifically designed to: address the discovered security issues, be efficient and simple to implement correctly, and of course, enable good compression of HTTP header metadata.
+
+For full details of the HPACK compression algorithm, see <https://tools.ietf.org/html/draft-ietf-httpbis-header-compression>.
+
+## Further reading:
+
+* [“HTTP/2”](https://hpbn.co/http2/) – The full article by Ilya Grigorik
+* [“Setting up HTTP/2”](https://surma.link/things/h2setup/) – How to set up HTTP/2 in different backends by Surma
```
Comment (petele) [inline src/content/en/fundamentals/performance/http2/index.md]: Should we also bring this across and publish it here?

**Comment by petele** [comment]
@jpmedley - can you (or an editor) take a swing through this please?

File: src/content/en/fundamentals/performance/http2/index.md (hunk 6 above)
Comment (surma) [inline src/content/en/fundamentals/performance/http2/index.md]: I’d argue no. It’s a timeline, nothing more. No real value in linking to outdated drafts, is there?

File: src/content/en/fundamentals/performance/http2/index.md (hunk 11 above)
Comment (surma) [inline src/content/en/fundamentals/performance/http2/index.md]: Yes, one of my next tasks. I’ll update the link here when with that PR.

**Comment by petele** [comment]
Please assign back to me once you've completed the updates.

**Comment by petele** [comment]
@jpmedley - assigned to you for edit. when done please assign back to me or Surma as appropriate.

File: src/content/en/fundamentals/performance/http2/index.md [hunk 12]
```diff
@@ -0,0 +1,568 @@
...
+* Pushed resources can be prioritized by the server
+* Pushed resources can be declined by the client
+
+### PUSH_PROMISE 101
+
+All server push streams are initiated via PUSH_PROMISE frames, which signal the
+server’s intent to push the described resources to the client and need to be
+delivered ahead of the response data that requests the pushed resources. This
+delivery order is critical: the client needs to know which resources the server
+intends to push to avoid creating own and duplicate requests for these
```
Comment (jpmedley) [inline src/content/en/fundamentals/performance/http2/index.md]: What's an 'own and duplicate request'?

File: src/content/en/fundamentals/performance/http2/index.md [hunk 13]
```diff
@@ -0,0 +1,568 @@
...
+
+Huffman coding allows the individual values to be compressed when transferred,
+and the indexed list of previously transferred values allows us to encode
+duplicate values by transferring index values that can be used to efficiently
+look up and reconstruct the full header keys and values.
+
+![](images/header_compression01.svg)
+
+As one further optimization, the HPACK compression context consists of a static
+and dynamic tables: the static table is defined in the specification and
```
Comment (jpmedley) [inline src/content/en/fundamentals/performance/http2/index.md]: 'a static and dynamic tables'

Not sure if this is a typo or intention. Are saying there is one static table and many dynamic tables or should the 'a' be removed.

File: src/content/en/fundamentals/performance/http2/index.md [hunk 14]
```diff
@@ -0,0 +1,571 @@
+project_path: /web/_project.yaml
+book_path: /web/fundamentals/_book.yaml
+description: HTTP/2 (or h2) is a binary protocol that brings push, multiplexing streams and frame control to the web.
+
+{# wf_updated_on: 2016-09-29 #}
+{# wf_published_on: 2016-09-29 #}
+
+# Introduction to HTTP/2 {: .page-title }
+
+{% include "web/_shared/contributors/surma.html" %}
```
Comment (PaulKinlan) [inline src/content/en/fundamentals/performance/http2/index.md]: It might be good to include Ilya as well....

**Comment by PaulKinlan** [comment]
@petele  Just did a read through it looks good. Only added one comment that we could fix post-merge.

File: src/content/en/fundamentals/performance/http2/index.md (hunk 14 above)
Comment (surma) [inline src/content/en/fundamentals/performance/http2/index.md]: LOL, totally right. I had that in the previous version... Sorry!

**Comment by igrigorik** [comment]
@surma @petele late to the game here.. Per O'Reilly policy, we need a callout/disclaimer at the top noting that this section is an excerpt from HPBN ~~: 

> "The following content is an excerpt from [High Performance Browser Networking](http://shop.oreilly.com/product/0636920028048.do) (O'Reilly, Ilya Grigorik). For full version and related content, see [hpbn.co](https://hpbn.co)"

Or some such..

**Comment by surma** [comment]
Will fix

**Pull request merged**